webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPageActions.js

```diff
@@ -0,0 +1,88 @@
+import history from '../../../history';
+import { API } from '../../../redux/API';
+import { translate as __ } from '../../../common/I18n';
+
+import { deepPropsToCamelCase } from '../../../common/helpers';
+import { buildQuery } from './ModelsPageHelpers';
+
+import {
+  MODELS_PAGE_DATA_RESOLVED,
+  MODELS_PAGE_DATA_FAILED,
+  MODELS_PAGE_CLEAR_ERROR,
+  MODELS_PAGE_SHOW_LOADING,
+  MODELS_API_PATH,
+  MODELS_PATH,
+} from '../constants';
+
+import { selectHasError } from './ModelsPageSelectors';
+
+import { stringifyParams, getParams } from '../../../common/urlHelpers';
+
+export const initializeModels = () => dispatch => {
+  const params = getParams();
+  dispatch(fetchModels(params));
+  if (!history.action === 'POP') {
+    history.replace({
+      pathname: MODELS_PATH,
+      search: stringifyParams(params),
+    });
+  }
+};
+
+export const fetchModels = (
+  { page, perPage, searchQuery, sort },
+  url = MODELS_API_PATH
+) => async (dispatch, getState) => {
+  dispatch({ type: MODELS_PAGE_SHOW_LOADING });
+
+  if (selectHasError(getState())) {
+    dispatch({ type: MODELS_PAGE_CLEAR_ERROR });
+  }
+
+  try {
+    const sortString =
+      sort && Object.keys(sort).length > 0 ? `${sort.by} ${sort.order}` : '';
+
+    const { data } = await API.get(
+      url,
+      {},
+      {
+        page,
+        per_page: perPage,
+        search: searchQuery,
+        order: sortString,
+      }
+    );
+
+    const transformedResponse = deepPropsToCamelCase(data);
+
+    dispatch({
+      type: MODELS_PAGE_DATA_RESOLVED,
+      payload: {
+        ...transformedResponse,
+        hasData: transformedResponse.subtotal > 0,
+        isLoading: false,
+      },
+    });
+  } catch (error) {
+    dispatch({
+      type: MODELS_PAGE_DATA_FAILED,
+      payload: {
+        message: {
+          type: 'error',
+          text: `${error.response.status} ${__(error.response.statusText)}`,
+        },
+        isLoading: false,
+      },
+    });
+  }
+};
+
+export const fetchAndPush = (params = {}) => (dispatch, getState) => {
```

Comment:
Can't we do only push and do the fetch on location change event?

Reply:
I am not sure what you mean, this follows the convention established by the audits page.